### PR TITLE
Fixes the ip address to namespace cache when pods reuse addresses

### PR DESCRIPTION
**Pull request by knobunc**
There's a cache that tracks which namespace IP addresses belong to.  However, the cache logic was not considering that Failed or Complete pods (i.e. not running) could share an IP address with a Running pod.  so the cache code needed to be changed to only add items to the cache when a pod was Running or Pending.  We also track the pod UID to make sure that when we delete, we are deleting the entry for the pod that was running.  So that if a stopped pod is edited, that entry doesn't end up in the cache.

This is reproducible by setting the networkConfig hostSubnetLength to 3 in the master and node config on your cluster.  Then create two namespaces, ns1 and ns2, a service in ns2 that has a selector that matches both pods then create a pod that fails in one (have it run "sleep 5; false", and set the restartPolicy to Never.  Once it has failed, create a pod that runs correctly in the other (hello-openshift is fine).  Then curling the service URL should work.  Remove the failed pod and the curl will then fail.

The problem is that the cache sees the failed pod removal and removes the ip from the ip->namespace cache.  But it should not have been removed because the running pod put the entry in.

For bug:
  https://bugzilla.redhat.com/show_bug.cgi?id=132294

Also-Authored-By: sross@redhat.com

### CODE REVIEW

**Comment by knobunc** [comment]
@sross PTAL

**Comment by DirectXMan12** [comment]
@knobunc :+1: thanks!

**Comment by knobunc** [comment]
@danwinship @pravisankar PTAL

**Comment by danwinship** [comment]
> There's a cache that tracks which namespace IP addresses belong to. However, the cache logic was not considering that Failed or Complete pods (i.e. not running) could share an IP address with a Running pod.

I still feel like *this* is the bug, and we ought to at least file a bug against kubernetes... (even while working around the problem here)

File: plugins/osdn/registry.go [hunk 1]
```diff
@@ -666,3 +671,39 @@ EndpointLoop:
 
 	registry.baseEndpointsHandler.OnEndpointsUpdate(filteredEndpoints)
 }
+
+func (registry *Registry) trackPod(pod *kapi.Pod) {
+	if pod.Status.PodIP != "" {
```
Comment (danwinship) [inline plugins/osdn/registry.go]: could simplify this by just

    if pod.Status.PodIP == "" {
        return
    }

File: plugins/osdn/registry.go [hunk 2]
```diff
@@ -666,3 +671,39 @@ EndpointLoop:
...
+			// we need to add it to our IP -> namespace tracker.  There _should_ be no
+			// other entries for the IP if we caught all of the right messages, so warn
+			// if we see one, but clobber it anyway since the IPAM
+			// should ensure that each IP is uniquely assigned to a pod (when running)
+			if ok && info.uid != pod.UID {
+				log.Warningf("IP '%s' was marked as used by namespace '%s' (pod '%s')... updating to namespace '%s' (pod '%s')",
+					pod.Status.PodIP, info.namespace, info.uid, pod.ObjectMeta.Namespace, pod.UID)
+			}
+
+			registry.namespaceOfPodIP[pod.Status.PodIP] = podInfo{namespace: pod.ObjectMeta.Namespace, uid: pod.UID}
```
Comment (danwinship) [inline plugins/osdn/registry.go]: maybe just store the *kapi.Pod itself for simplicity? Also, "namespaceOfPodIP" is no longer an accurate description of the data...

**Comment by danwinship** [comment]
basically looks good

File: plugins/osdn/registry.go (hunk 2 above)
Comment (pravisankar) [inline plugins/osdn/registry.go]: Yes, change 'namespaceOfPodIP' to 'podInfoMap' or some other better name

**Comment by pravisankar** [comment]
LGTM after danwinship's comments are addressed

**Comment by knobunc** [comment]
@danwinship @pravisankar @DirectXMan12 PTAL

**Comment by knobunc** [comment]
@danwinship consensus on slack #kubernetes-dev (@ncdc et al) was that multiple pods can have the same podIP as long as only one is running at a time.

**Comment by pravisankar** [comment]
lgtm

**Comment by smarterclayton** [comment]
Was this an existing issue that no one encountered?  Or did we introduce it with other changes?  If so, what release was this introduced?

**Comment by danwinship** [comment]
It doesn't seem to be caused by any change in openshift-sdn, but it also seems like people would have noticed this bug sooner if it had always been around. So either it's a regression caused by changes outside of openshift-sdn, or else it's just a sign that not many people were using multitenant before.

**Pull request merged**